File: bumperbot_cpp_pkg/src/simple_tf_kinematics.cpp
#include "bumperbot_cpp_pkg/simple_tf_kinematics.hpp"

using namespace std::chrono_literals;
using std::placeholders::_1;
using std::placeholders::_2;

SimpleTfKinematics::SimpleTfKinematics(const std::string &name) : Node(name)
{
    static_tf_broadcaster_ = std::make_shared<tf2_ros::StaticTransformBroadcaster>(this);
    dynamic_tf_broadcaster_ = std::make_shared<tf2_ros::StaticTransformBroadcaster>(this);
    tf_buffer_ = std::make_unique<tf2_ros::Buffer>(get_clock());
    tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);

    static_tf_msg_.header.frame_id = "bumperbot_base";
    static_tf_msg_.child_frame_id = "bumperbot_top";
    static_tf_msg_.header.stamp = get_clock()->now();
    static_tf_msg_.transform.translation.x = 0.0;
    static_tf_msg_.transform.translation.y = 0.0;
    static_tf_msg_.transform.translation.z = 0.3;
    static_tf_msg_.transform.rotation.x = 0.0;
    static_tf_msg_.transform.rotation.y = 0.0;
    static_tf_msg_.transform.rotation.z = 0.0;
    static_tf_msg_.transform.rotation.w = 1.0;

    static_tf_broadcaster_->sendTransform(static_tf_msg_);

    RCLCPP_INFO_STREAM(get_logger(), "Publishing static transform between" << static_tf_msg_.header.frame_id << "and" << static_tf_msg_.child_frame_id);
    timer_ = create_wall_timer(0.1s, std::bind(&SimpleTfKinematics::timer_callback, this));

    get_transform_srv_ = create_service<bumperbot_msgs::srv::GetTransform>("get_transform", std::bind(&SimpleTfKinematics::get_tf_callback, this, _1, _2));
}

void SimpleTfKinematics::timer_callback()
{
    dynamic_tf_msg_.header.stamp = get_clock()->now();
    dynamic_tf_msg_.header.frame_id = "odom";
    dynamic_tf_msg_.child_frame_id = "bumperbot_base";
    dynamic_tf_msg_.transform.translation.x = last_x_ + x_increment_;
    dynamic_tf_msg_.transform.translation.y = 0.0;
    dynamic_tf_msg_.transform.translation.z = 0.0;
    dynamic_tf_msg_.transform.rotation.x = 0.0;
    dynamic_tf_msg_.transform.rotation.y = 0.0;
    dynamic_tf_msg_.transform.rotation.z = 0.0;
    dynamic_tf_msg_.transform.rotation.w = 1.0;
    dynamic_tf_broadcaster_->sendTransform(dynamic_tf_msg_);
    last_x_ = dynamic_tf_msg_.transform.translation.x;
}

bool SimpleTfKinematics::get_tf_callback(const std::shared_ptr<bumperbot_msgs::srv::GetTransform::Request> request,
                                         const std::shared_ptr<bumperbot_msgs::srv::GetTransform::Response> response)
{
    RCLCPP_INFO_STREAM(get_logger(), "Requested transform between " << request->frame_id << " and " << request->child_frame_id);
    geometry_msgs::msg::TransformStamped requested_transform;
    try
    {
        requested_transform = tf_buffer_->lookupTransform(request->frame_id, request->child_frame_id, tf2::TimePointZero);
    }
    catch (tf2::TransformException &ex)
    {
        RCLCPP_ERROR_STREAM(get_logger(), "Error in accessing transform between " << request->frame_id << "and" << request->child_frame_id << ": " << ex.what());
        response->success = false;
        return true;
    }
    response->transform = requested_transform;
    response->success = true;
    return true;
}

int main(int argc, char *argv[])
{

    rclcpp::init(argc, argv);
    auto node = std::make_shared<SimpleTfKinematics>("simple_tf_kinematics");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: bumperbot_py_pkg/bumperbot_py_pkg/simple_tf_kinematics.py
import rclpy
from rclpy.node import Node
from tf2_ros.static_transform_broadcaster import StaticTransformBroadcaster
from tf2_ros import TransformBroadcaster, TransformException
from tf2_ros.buffer import Buffer
from tf2_ros.transform_listener import TransformListener
from geometry_msgs.msg import TransformStamped
from bumperbot_msgs.srv import GetTransform


class SimpleTfKinematics(Node):
    def __init__(self):
        super().__init__("simple_tf_kinematics")

        self.static_tf_broadcaster_ = StaticTransformBroadcaster(self)
        self.dynamic_tf_broadcaster_ = TransformBroadcaster(self)

        # dynamic
        self.dynamic_tf_msg_ = TransformStamped()
        self.x_increment_ = 0.05
        self.last_x_ = 0.0
        self.tf_buffer_ = Buffer()
        self.tf_listener_ = TransformListener(self.tf_buffer_, self)

        # static
        self.static_tf_msg_ = TransformStamped()
        self.static_tf_msg_.header.stamp = self.get_clock().now().to_msg()
        self.static_tf_msg_.header.frame_id = "bumperbot_base"
        self.static_tf_msg_.child_frame_id = "bumperbot_top"
        self.static_tf_msg_.transform.translation.x = 0.0
        self.static_tf_msg_.transform.translation.y = 0.0
        self.static_tf_msg_.transform.translation.z = 0.3
        self.static_tf_msg_.transform.rotation.x = 0.0
        self.static_tf_msg_.transform.rotation.y = 0.0
        self.static_tf_msg_.transform.rotation.z = 0.0
        self.static_tf_msg_.transform.rotation.w = 1.0

        self.static_tf_broadcaster_.sendTransform(self.static_tf_msg_)

        self.get_logger().info(
            f"Publishing static fransform between {self.static_tf_msg_.header.frame_id} and {self.static_tf_msg_.child_frame_id}")

        self.timer_ = self.create_timer(0.1, self.timer_callback)

        self.get_tf_srv_ = self.create_service(
            GetTransform, "get_transform", self.get_tf_callback)

    def timer_callback(self):
        self.dynamic_tf_msg_.header.stamp = self.get_clock().now().to_msg()
        self.dynamic_tf_msg_.header.frame_id = "odom"
        self.dynamic_tf_msg_.child_frame_id = "bumperbot_base"
        self.dynamic_tf_msg_.transform.translation.x = self.last_x_ + self.x_increment_
        self.dynamic_tf_msg_.transform.translation.y = 0.0
        self.dynamic_tf_msg_.transform.translation.z = 0.0
        self.dynamic_tf_msg_.transform.rotation.x = 0.0
        self.dynamic_tf_msg_.transform.rotation.y = 0.0
        self.dynamic_tf_msg_.transform.rotation.z = 0.0
        self.dynamic_tf_msg_.transform.rotation.w = 1.0

        self.dynamic_tf_broadcaster_.sendTransform(self.dynamic_tf_msg_)

        self.last_x_ = self.dynamic_tf_msg_.transform.translation.x

    def get_tf_callback(self, request, response):
        self.get_logger().info(
            f"Requested transform between {request.frame_id}, and {request.child_frame_id}")
        requested_transform = TransformStamped()
        try:
            requested_transform = self.tf_buffer_.lookup_transform(
                request.frame_id, request.child_frame_id, rclpy.time.Time())

        except TransformException as e:
            self.get_logger.error(
                f"Error occured while accessing fransform between {self.static_tf_msg_.header.frame_id} and {self.static_tf_msg_.child_frame_id}")
            response.success = False

        response.transform = requested_transform
        response.success = True
        return response


def main():
    rclpy.init()
    node = SimpleTfKinematics()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()
